# 异常体系

## 1. 异常继承体系

​	Java中异常体系的核心类是Throwable，它有两个子类Error和Exception。

- Error代表一些非常严重的错误。我们一般不必特意在代码中处理他们。
- Exception相当于一些小错误。可以用来提示我们出现了什么问题。我们后面主要讲的就是Exception。



异常主要分两种：

- 运行时异常（编译期间不会去做检查，不需要在代码中做预处理）

  运行时异常都是RuntimeException的子类，

  例如：NullPointerException, ArrayIndexOutOfBoundsException

- 编译时异常（编译时就会做检查，如果一段代码中可能出现编译时异常必须在代码中做预处理）

  1. 非继承自RuntimeException

  2. Exception的子类

  例如：FileNotFoundException

​	





## 2. 异常处理

### 2.1 throws声明抛出异常

​	有些时候我们需要把异常抛出，在适当的地方去处理异常。这个时候就可以使用throws抛出异常，把异常交给方法的调用者处理。

格式：

​		` public static void XX() throws XXException`

​		在方法声明出加上 throws 异常类型.如果有多个异常用逗号分隔

示例：

抛出一个异常

~~~~java
    public static void test() throws FileNotFoundException {
        FileInputStream fis = new FileInputStream("a.txt");
    }
~~~~

抛出多个异常

~~~~java
    public static void test() throws FileNotFoundException, ParseException {
        FileInputStream fis = new FileInputStream("a.txt");
    }
~~~~



### 2.2 try...catch抓取异常

​		我们可以对可能出现异常的代码提前做好准备。这个时候就可以使用到try...catch了。

#### 2.2.1 格式

~~~~java
	try{
        //可能出现异常的代码
    }catch(异常类型 标识符){
        //对异常做相应的处理
    }
~~~~

示例：

~~~~java
    public static void main(String[] args) {
        try {
            test();
        } catch (FileNotFoundException e) {
            e.printStackTrace();//打印异常信息
        }
    }
~~~~



#### 2.2.2 执行流程

​		如果try中的代码真的出现了异常，并且异常类型和cathch中的异常类型能匹配上。这个时候就会进入指定的catch块中执行，而异常对象也会赋值给我们在catch中定义的异常变量接收。

​		如果没有异常try中代码执行完就去执行try..catch之后的代码。

#### 2.2.3 注意事项

如果可能有多种异常，有多种处理方式。

①可以选择使用多个catch。

例如：

~~~~java
    public static void main(String[] args) {
        try {
            test();
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (ParseException e) {
            e.printStackTrace();
        }
    }
~~~~

②直接catch一个父异常类型

~~~~java
    public static void main(String[] args) {
        try {
            test();
        } catch (Exception e) {
            e.printStackTrace();
        } 
    }
~~~~





## 3. 自定义异常

​		我们也可以自定义异常类。如果要定义运行时异常只要继承RuntimeException,定义构造方法即可。如果是定义编译时异常改成继承Exception。

示例：

~~~~java
//自定义运行时异常
public class MyRuntimeException extends RuntimeException {
    public MyRuntimeException() {
    }

    public MyRuntimeException(String message) {
        super(message);
    }
}
~~~~

~~~~java
//自定义编译时异常
public class TooSmallException extends Exception {
    public TooSmallException() {
    }

    public TooSmallException(String message) {
        super(message);
    }
}

~~~~



我们可以去使用下自定义的异常。可以在代码中使用throw抛出异常对象。

~~~~java
    public static void playGame(int age) throws TooSmallException {
        if(age<18){
            throw new TooSmallException("您的年龄过小不适合玩本游戏。");
        }
    }
~~~~

因为自定义的TooSmallException是一个编译时异常，所以必须对异常进行预处理。如果是适合运行时异常不做预处理也没有问题。



## 4. 异常的作用

1. 异常可以帮助我们知道具体的错误原因。
2. 异常可以让我们在方法调用过程中出现问题的时候，把具体的问题反馈到方法调用处。

